### PR TITLE
 Added config map reporter to synchronize config gateway

### PR DESCRIPTION
**Pull request by jshaofuturewei**
The PR is to synchronize a cluster gateway host ips as neighbors with its registered cluster.

Each cluster might have a gateway. We need all the clusters know their neighbor for vpc broadcast. Therefore, we update the cluster gateway config map from
```bash
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-gateway-config
  namespace: default
data:
  gateway_host_ip: 1.0.0.2
```
To 
```bash
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-gateway-config
  namespace: default
data:
  gateway_host_ip: 1.0.0.2
  gateway_neighbors: 1.0.0.1, 1.0.0.3, ...
```
In order to make it automatically, when a cluster is registered as sub edgecluster in fornax, we create a config map reporter to synchronize its configured gateway host ip to the cluster it registers. Please check the following user senarios.

1. Set up two fornax clusters node-a and node-c
2. Start cloudhub in node-a
3. Start edgehub in node-c
4. Verify that node-c is registered as sub edge cluster in node-a
```bash
root@node-a:~/go/src/github.com/fornax# kubectl get edgeclusters
NAME     LASTHEARBEAT   HEALTHSTATUS   SUBEDGECLUSTERS   RECEIVED_MISSIONS       MATCHED_MISSIONS
node-c   3s             healthy                          ["deployment-to-all"]   ["deployment-to-all"]
```
5 Apply the following cluster gateway map to a cluster node-a
```bash
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-gateway-config
  namespace: default
data:
  gateway_host_ip: 1.0.0.2
```
6 Apply the following cluster gateway map to a cluster node-c
```bash
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-gateway-config
  namespace: default
data:
  gateway_host_ip: 1.0.0.1
```
7. Check the node-a cloudcore.logs and find the following sync messages
```bash
I0318 18:44:31.363701    2435 upstream.go:188] dispatch message content: {Header:{ID:e57e85c2-3213-4427-8735-039f76f99ca7 ParentID: Timestamp:1647629071349 ResourceVersion: Sync:false} Router:{Source:clusterd Group:meta Operation:update Resource:node/node-c/default/configmap/node-c} Content:map[data:map[gateway_host_ip:1.0.0.1] metadata:map[clusterName:node-c creationTimestamp:<nil>]]}
```
8. Verify that the node-c gateway 1.0,0.1 is added to node-a configmap as neighbors
```bash
 kubectl get configmap cluster-gateway-config -o=jsonpath='{.data}'
{"gateway_host_ip":"1.0.0.2","gateway_neighbors":"1.0.0.1"}
```

### CODE REVIEW

File: cloud/pkg/edgecontroller/controller/downstream.go [hunk 1]
```diff
@@ -133,6 +134,25 @@ func (dc *DownstreamController) syncConfigMap() {
 				continue // continue to next select
 			}
 
+			// It is used to inform the local cluster gateway changes to its neighbors
+			if configMap.Name == "cluster-gateway-config" {
```
Comment (pdgetrf) [inline cloud/pkg/edgecontroller/controller/downstream.go:138]: add this string to a variable to be shared with other places

File: cloud/pkg/edgecontroller/controller/downstream.go [hunk 2]
```diff
@@ -144,11 +164,11 @@ func (dc *DownstreamController) syncConfigMap() {
 					klog.Warningf("build message resource failed with error: %s", err)
 					continue
 				}
+
```
Comment (pdgetrf) [inline cloud/pkg/edgecontroller/controller/downstream.go:167]: nit: extra space

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 3]
```diff
@@ -1254,5 +1263,72 @@ func NewUpstreamController(factory k8sinformer.SharedInformerFactory) (*Upstream
...
 }
+
+func (uc *UpstreamController) updateClusterGatewayNeighbors() {
+	for {
+		select {
+		case <-beehiveContext.Done():
+			klog.Warning("stop updateClusterGatewayNeighbors")
+			return
+		case msg := <-uc.clusterGatewayChan:
+			klog.V(5).Infof("message: %s, operation is: %s, and resource is %s", msg.GetID(), msg.GetOperation(), msg.GetResource())
```
Comment (pdgetrf) [inline cloud/pkg/edgecontroller/controller/upstream.go:1277]: probably remove "message:" in such logs. it doesn't provide additional info.

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 4]
```diff
@@ -1254,5 +1263,72 @@ func NewUpstreamController(factory k8sinformer.SharedInformerFactory) (*Upstream
...
+
+			namespace, err := messagelayer.GetNamespace(msg)
+			if err != nil {
+				klog.Warningf("message: %s process failure, get namespace failed with error: %s", msg.GetID(), err)
```
Comment (pdgetrf) [inline cloud/pkg/edgecontroller/controller/upstream.go:1281]: here too.

File: edge/pkg/clusterd/cluster_gateway_reporter.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+type ClusterGatewayReporter struct {
+	clusterd                     *clusterd
+	clusterGatewayUpdateInterval time.Duration
+}
+
+func NewClusterGatewayReporter(c *clusterd) *ClusterGatewayReporter {
+	return &ClusterGatewayReporter{
+		clusterd:                     c,
+		clusterGatewayUpdateInterval: 60 * time.Second,
```
Comment (pdgetrf) [inline edge/pkg/clusterd/cluster_gateway_reporter.go:40]: 60 * time.Second --> 1 time.Minute?

File: edge/pkg/clusterd/cluster_gateway_reporter.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	if err != nil {
+		klog.Errorf("update cluster gateway config map failed, error: %v", err)
+		return err
+	}
+	return nil
+}
+
+func (reporter *ClusterGatewayReporter) syncClusterGatewayConfigMap() {
+	if err := reporter.updateClusterGatewayConfigMap(); err != nil {
+		klog.Errorf("Unable to update cluster gateway config maps: %v", err)
```
Comment (pdgetrf) [inline edge/pkg/clusterd/cluster_gateway_reporter.go:63]: first character lower case to be consistent with other logs

File: edge/pkg/clusterd/cluster_gateway_reporter.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func (reporter *ClusterGatewayReporter) updateClusterGatewayConfigMap() error {
+	clusterGatewayHostIP, err := GetClusterGatewayHostIP()
+	if err != nil {
+		return err
+	}
+	configMap := &corev1.ConfigMap{}
+	configMap.Data = make(map[string]string)
+	configMap.Data["gateway_host_ip"] = clusterGatewayHostIP
```
Comment (pdgetrf) [inline edge/pkg/clusterd/cluster_gateway_reporter.go:51]: use a variable for "gateway_host_ip"

**Review by pdgetrf — APPROVED** [review]
LGTM

a few minor suggestions.

**Comment by pdgetrf** [comment]
also, seems the TEST is failing. I've click the re-run and let's see how it goes this time.

**Comment by pdgetrf** [comment]
> also, seems the TEST is failing. I've click the re-run and let's see how it goes this time.

tests have just passed.

File: edge/pkg/clusterd/cluster_gateway_reporter.go (hunk 7 above)
Comment (jshaofuturewei) [reply edge/pkg/clusterd/cluster_gateway_reporter.go:51]: Updated

File: edge/pkg/clusterd/cluster_gateway_reporter.go (hunk 6 above)
Comment (jshaofuturewei) [reply edge/pkg/clusterd/cluster_gateway_reporter.go:63]: Updated

File: edge/pkg/clusterd/cluster_gateway_reporter.go (hunk 5 above)
Comment (jshaofuturewei) [reply edge/pkg/clusterd/cluster_gateway_reporter.go:40]: Updated

File: cloud/pkg/edgecontroller/controller/downstream.go (hunk 1 above)
Comment (jshaofuturewei) [reply cloud/pkg/edgecontroller/controller/downstream.go]: Added

File: cloud/pkg/edgecontroller/controller/downstream.go (hunk 2 above)
Comment (jshaofuturewei) [reply cloud/pkg/edgecontroller/controller/downstream.go]: Removed

File: cloud/pkg/edgecontroller/controller/upstream.go (hunk 3 above)
Comment (jshaofuturewei) [reply cloud/pkg/edgecontroller/controller/upstream.go]: Removed

File: cloud/pkg/edgecontroller/controller/upstream.go (hunk 4 above)
Comment (jshaofuturewei) [reply cloud/pkg/edgecontroller/controller/upstream.go]: Removed

**Comment by pdgetrf** [comment]
thanks for the quick update. there's some lint errors.

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 8]
```diff
@@ -1254,5 +1267,64 @@ func NewUpstreamController(factory k8sinformer.SharedInformerFactory) (*Upstream
...
+				neighbors = configMap.Data["gateway_neighbors"]
+			}
+			if updatedNeighbors, updated, err := util.GetUpdatedClusterGatewayNeighbors(neighborName, neighborHostIP, neighbors); updated && err == nil {
+				configMap.Data["gateway_neighbors"] = updatedNeighbors
+				updatedConfigMap, err := uc.kubeClient.CoreV1().ConfigMaps(namespace).Update(context.Background(), configMap, metaV1.UpdateOptions{})
+				if err != nil {
+					klog.Warningf("Failed to updated neighbors, namespace: %s, name: %s, err: %v", namespace, name, err)
+					continue
+				}
+				klog.V(3).Infof("upstream gateway configmap updated %v", updatedConfigMap)
```
Comment (pdgetrf) [inline cloud/pkg/edgecontroller/controller/upstream.go:1324]: some logs are starting with lower case letters while others are upper cases (line 1321 for example).

**Comment by pdgetrf** [comment]
@jshaofuturewei one little comment about mixing upper and lower case log messages. not a big issue. I'm okay merging it as is or if you prefer to fix it, that's okay too.

**Review by pdgetrf — APPROVED** [review]
LGTM

File: cloud/pkg/edgecontroller/controller/upstream.go (hunk 8 above)
Comment (jshaofuturewei) [reply cloud/pkg/edgecontroller/controller/upstream.go:1324]: Updated

**Comment by jshaofuturewei** [comment]
> @jshaofuturewei one little comment about mixing upper and lower case log messages. not a big issue. I'm okay merging it as is or if you prefer to fix it, that's okay too.

Most of messages are copied from existing codes. I have updated them to lowercases noew.

**Comment by pdgetrf** [comment]
> > @jshaofuturewei one little comment about mixing upper and lower case log messages. not a big issue. I'm okay merging it as is or if you prefer to fix it, that's okay too.
> 
> Most of messages are copied from existing codes. I have updated them to lowercases noew.

alright. thanks. I'll merge it if there's no more update after the CI runs.

**Pull request merged**